index.js

```diff
@@ -1,395 +1,157 @@
-}
-
-function compile (schema, opts) {
-  if (!opts) opts = {}
-
-  const unsafe = bool(opts.unsafe, false)
-  const validate = bool(opts.validate, !unsafe)
-  const optional = bool(opts.optional, true)
-  const unescapeStrings = bool(opts.unescapeStrings, true)
-  const validateKeys = bool(opts.validateKeys, !unsafe)
-  const validateLength = bool(opts.validateLength, !unsafe)
-  const buffer = !!opts.buffer
-
-  var tick = 0
-  var incs = 0
-  var absVar = ''
-  var declaredPtr = false
+const genfun = require('generate-function')
 
-  const constants = {endNumber, indexOfString, indexOfKey, endNumberBuffer, endNumberString, stringUnescape}
+module.exports = function (schema) {
   const gen = genfun()
 
-  if (typeof schema !== 'object') throw new Error('Only object or array schemas supported')
+  const constants = {}
 
-  gen('function parse (s, b) {')
-  compileType(null, schema, gen)
-  gen('}')
-
-  return gen.toFunction(constants)
+  var tick = 0
 
-  function inc (n, abs) {
-    if (abs) {
-      absVar = abs
-      incs = n
-    } else {
-      incs += n
-    }
+  function tmp () {
+    return 'tmp' + (tick++)
   }
 
-  function flush () {
-    const decl = declaredPtr ? 'ptr' : 'var ptr'
-    if (absVar) {
-      gen('%s = %s + %d', decl, absVar, incs)
-      incs = 0
-      absVar = ''
-    } else if (incs) {
-      const plus = declaredPtr ? '+=' : '='
-      gen('%s %s %d', decl, plus, incs)
-      incs = 0
-    } else if (!declaredPtr) {
-      gen('%s = 0', decl)
+  function def (type) {
+    if (typeof type === 'string') {
+      switch (type) {
+        case 'string': return '""'
+        case 'number': return '0'
+        case 'boolean':
+        case 'bool': return 'false'
+        case 'array': return '[]'
+        case 'object': return '{}'
+        default: throw new Error('Not default type available')
+      }
     }
-    declaredPtr = true
-  }
-
-  function tmp () {
-    return 'tmp' + (tick++)
+    return 'undefined'
   }
 
-  function defaultObj (obj, fields) {
+  function defaultObj (schema, fields) {
     for (var j = 0; j < fields.length; j++) {
       const name = fields[j]
       const sep = j < fields.length - 1 ? ',' : ''
-      const child = obj[name]
-      if (isObject(child) && !optional) {
-        gen('%s: {', name)
-        defaultObj(child, Object.keys(child))
-        gen('}%s', sep)
-      } else if (Array.isArray(child) && !optional) {
-        gen('%s: []%s', name, sep)
-      } else {
-        gen('%s: %s%s', name, def(child), sep)
-      }
+      const child = schema.properties[name]
+      gen('%s: %s%s', name, def(child.type), sep)
     }
   }
 
-  function ch (i) {
-    return buffer ? 'b[' + i + ']' : 's.charCodeAt(' + i + ')'
+  function generateString (fieldName, startValueVar) {
+    gen(`if (s.charCodeAt(ptr++) !== 34) throw new Error("Cannot parse object")`)
+    gen('%s = ptr', startValueVar)
+    gen(`while (s.charCodeAt(++ptr) !== 34) ;`)
```

Comment:
How does it handle the escape character? Example: `"foo\"bar"`

Reply:
Yeah, you are right. I'll update asap